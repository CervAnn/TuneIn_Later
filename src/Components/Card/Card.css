.Card {
    height: 300px;
    width: 300px;
    background-color: bisque;
    display: flex;
    align-items: center;
    justify-content: space-evenly;
    flex-direction: column;
    border: 2px solid black;
    margin-bottom: 10px;
}

img {
    object-fit: cover;
    height: 100px;
    width: 100px;
}