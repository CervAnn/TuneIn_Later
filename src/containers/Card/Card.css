.Card, .favoriteCard {
  height: 350px;
  width: 300px;
  background-color: #F4E697;
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  flex-direction: column;
  border: 4px solid #544061;
  margin-bottom: 10px;
  border-radius: 15px;
  opacity: .8;
}

img {
    object-fit: cover;
    height: 150px;
    width: 150px;
}

h2 {
  padding: 10px;
  align-self: auto;
}

button {
  background-color: transparent;
  background-position: center;
  border: none;
}

.favorite {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

#isFavorite, #notFavorite {
  height: auto;
  width: 15%;
}