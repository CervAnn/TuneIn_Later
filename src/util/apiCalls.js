export const fetchAlbums = queryArtist => {
  return fetch(
    `https://itunes.apple.com/search?term=${queryArtist.name}&entity=album`
  ).then(response => {
    if (!response.ok) {
      throw Error(
        "There was an issue retrieving your artist's albums. Please try again."
      );
    }
    return response.json();
  });
  // .catch(error => {
  //   throw Error ('There was an issue retrieving your artist\'s albums. Please try again.')
  // })
};

export const fetchUser = user => {
  const options = {
    method: "POST",
    body: JSON.stringify(user),
    headers: {
      "Content-Type": "application/json"
    }
  };
  return fetch("http://localhost:3001/api/v1/login", options).then(response => {
    if (!response.ok) {
      throw Error(
        "There was an issue retrieving your account information. Please try again."
      );
    }
    return response.json();
  });
};

export const addUser = user => {
  const options = {
    method: "POST",
    body: JSON.stringify(user),
    headers: {
      "Content-Type": "application/json"
    }
  };
  return fetch("http://localhost:3001/api/v1/users", options).then(response => {
    if (!response.ok) {
      throw Error("There was an issue creating your account.");
    }
    return response.json();
  });
};

export const fetchUserFavorites = id => {
  return fetch(`http://localhost:3001/api/v1/users/${id}/albumfavorites`).then(
    response => {
      if (!response.ok) {
        throw Error("There was an issue getting your favorites.");
      }
      return response.json();
    }
  );
};

export const postFavorite = (card, id) => {
  const options = {
    method: "POST",
    body: JSON.stringify(card),
    headers: {
      "Content-Type": "application/json"
    }
  };

  return fetch(
    `http://localhost:3001/api/v1/users/${id}/albumfavorites`,
    options
  ).then(response => {
    if (!response.ok) {
      throw Error("There was an issue adding your favorite.");
    }
    return response.json();
  });
};

export const deleteFavorite = (albumId, userId) => {
  const options = {
    method: "DELETE"
  };
  return fetch(
    `http://localhost:3001/api/v1/users/${userId}/albumfavorites/${albumId}`,
    options
  ).then(response => {
    if (!response.ok) {
      throw Error(
        "There was an issue deleting your favorite. You're stuck with it."
      );
    }
  });
};
