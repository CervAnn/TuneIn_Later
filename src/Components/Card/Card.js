import React from 'react'
import './Card.css'
import { addFavorite, invalidUser } from '../../actions/index'
import { connect } from 'react-redux';
import { postFavorite, deleteFavorite } from '../../util/apiCalls';

const Card = ({ album_id, artist_name, album_name, artwork_url, release_date, content_advisory_rating, primary_genre_name, user, addFavorite, invalidUser, isFavorite, favorites }) => {
	const isUserLoggedIn = (e) => {
		e.preventDefault()
		return user ? toggleFavorite() : invalidUser();
	}

	const toggleFavorite = () => {
		isFavorite = favorites.map(favorite => favorite.id).includes(album_id)
		if(isFavorite) {
			deleteFavorite(album_id)
		} else {
			const albumData = {album_id, artist_name, album_name, artwork_url, release_date, content_advisory_rating, primary_genre_name}
			postFavorite(albumData, user.id)
			.then(data => addFavorite(data))
			.catch(error => console.log(error))

		}
		isFavorite = !isFavorite;
	}

 	return (
		<article className="Card">
			<img src={artwork_url} alt="Album cover art"/>
			<h2>{album_name}</h2>
			<button onClick={(e) => isUserLoggedIn(e)}>Save For Later</button>
		</article>
	)
}
const mapStateToProps = (state) => ({
favorites: state.favorites
})

const mapDispatchToProps = dispatch => ({
	addFavorite: () => dispatch(addFavorite()),
	invalidUser: () => dispatch(invalidUser())
});

export default connect(null, mapDispatchToProps)(Card)
