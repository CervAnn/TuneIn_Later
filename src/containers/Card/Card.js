import React from 'react'
import './Card.css'
import { addFavorite, invalidUser, removeFavoriteFromStore } from '../../actions/index';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { postFavorite, deleteFavorite } from '../../util/apiCalls';

const Card = ({ album_id, artist_name, album_name, artwork_url, release_date, content_advisory_rating, primary_genre_name, user, addFavorite, invalidUser, favorites, removeFavoriteFromStore, routeType }) => {
	const isUserLoggedIn = (e) => {
		e.preventDefault()
		return user ? toggleFavorite() : invalidUser();
	}

	let isFavorite = favorites.map(favorite => favorite.album_id).includes(album_id);

	let cardClassName = isFavorite ? 'favoriteCard' : 'Card'

	const toggleFavorite = () => {
		if(isFavorite) {
			removeFavoriteFromStore(album_id);
			deleteFavorite(album_id, user.id);
		} else {
			const albumData = {album_id, artist_name, album_name, artwork_url, release_date, content_advisory_rating, primary_genre_name}
			postFavorite(albumData, user.id)
			.then(favorite => addFavorite(favorite))
			.catch(error => console.log(error))
		}
	}

 	return (
		<article className={cardClassName}>
				<Link to={`/${routeType}/${album_id}`}><img src={artwork_url} alt="Album cover art" /></Link>
			<h2>{album_name}</h2>
			<button onClick={(e) => isUserLoggedIn(e)}>Save For Later</button>
		</article>
	)
}
const mapStateToProps = (state) => ({
	favorites: state.favorites
})

const mapDispatchToProps = dispatch => ({
	addFavorite: (favorite) => dispatch(addFavorite(favorite)),
	invalidUser: () => dispatch(invalidUser()),
	removeFavoriteFromStore: (album_id) => dispatch(removeFavoriteFromStore(album_id))
});

export default connect(mapStateToProps	, mapDispatchToProps)(Card)
