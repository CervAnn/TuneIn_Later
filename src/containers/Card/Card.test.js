import React from 'react';
import { shallow } from 'enzyme';
import { Card, mapStateToProps, mapDispatchToProps } from './Card';
import { deleteFavorite, postFavorite } from '../../util/apiCalls';

describe('Card', () => {
  let albumWrapper, favoriteWrapper, newFavoriteWrapper, mockUser, mockAddFavorite, mockInvalidUser, mockFavorites, mockRemoveFavoriteFromStore;

  beforeEach(() => {
    mockUser = {
      id: 1,
      name: 'Alan',
      email: 'alan@turing.io'
    }
    mockAddFavorite = jest.fn();
    mockInvalidUser = jest.fn();
    mockFavorites = [
      {
      album_id: 626204707,
      artist_name: 'Beyoncé',
      album_name: '4 (Expanded Edition)',
      price: 11.99,
      artwork_url: 'https://is2-ssl.mzstatic.com/image/thumb/Music6/v4/17/84/3a/17843a6d-1f2b-7e1e-a39f-3ff865110993/source/100x100bb.jpg',
      release_date: '2013-12-13T08:00:00Z',
      content_advisory_rating: 'Explicit',
      primary_genre_name: 'Pop',
      user: mockUser,
      routeType: 'favorites'
      }
    ]
    mockRemoveFavoriteFromStore = jest.fn();
    albumWrapper = shallow(
    <Card 
      album_id={626204707}
      key={626204707}
      artist_name={'Beyoncé'}
      album_name={'4 (Expanded Edition)'}
      price={11.99}
      artwork_url={'https://is2-ssl.mzstatic.com/image/thumb/Music6/v4/17/84/3a/17843a6d-1f2b-7e1e-a39f-3ff865110993/source/100x100bb.jpg'}
      release_date={'2013-12-13T08:00:00Z'}
      content_advisory_rating={'Explicit'}
      primary_genre_name={'Pop'}
      user={mockUser}
      addFavorite={mockAddFavorite}
      invalidUser={mockInvalidUser}
      favorites={mockFavorites}
      removeFavoriteFromStore={mockRemoveFavoriteFromStore}
      routeType={'albums'}
    />)
    favoriteWrapper = shallow(
    <Card
      album_id={626204707}
      key={626204707}
      artist_name={'Beyoncé'}
      album_name={'4 (Expanded Edition)'}
      price={11.99}
      artwork_url={'https://is2-ssl.mzstatic.com/image/thumb/Music6/v4/17/84/3a/17843a6d-1f2b-7e1e-a39f-3ff865110993/source/100x100bb.jpg'}
      release_date={'2013-12-13T08:00:00Z'}
      content_advisory_rating={'Explicit'}
      primary_genre_name={'Pop'}
      user={mockUser}
      addFavorite={mockAddFavorite}
      invalidUser={mockInvalidUser}
      favorites={mockFavorites}
      removeFavoriteFromStore={mockRemoveFavoriteFromStore}
      routeType={'favorites'}
    />)
    newFavoriteWrapper = shallow(
    <Card
      album_id={626204708}
      key={626204707}
      artist_name={'Beyoncé'}
      album_name={'4 (Expanded Edition)'}
      price={11.99}
      artwork_url={'https://is2-ssl.mzstatic.com/image/thumb/Music6/v4/17/84/3a/17843a6d-1f2b-7e1e-a39f-3ff865110993/source/100x100bb.jpg'}
      release_date={'2013-12-13T08:00:00Z'}
      content_advisory_rating={'Explicit'}
      primary_genre_name={'Pop'}
      user={mockUser}
      addFavorite={mockAddFavorite}
      invalidUser={mockInvalidUser}
      favorites={mockFavorites}
      removeFavoriteFromStore={mockRemoveFavoriteFromStore}
      routeType={'favorites'}
    />)
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(albumWrapper).toMatchSnapshot();
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(favoriteWrapper).toMatchSnapshot();
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(newFavoriteWrapper).toMatchSnapshot();
  });

  it('should call deleteFavorite when isFavorite is true', () => {

    favoriteWrapper.toggleFavorite = jest.fn()
    favoriteWrapper.toggleFavorite();

    expect(deleteFavorite).toHaveBeenCalled();
  });

  it('should call postFavorite when isFavorite is false', () => {
    newFavoriteWrapper.toggleFavorite = jest.fn()
    newFavoriteWrapper.toggleFavorite();

    expect(postFavorite).toHaveBeenCalled();
  });

  it('should call isUserLoggedIn on a button click', () => {
    
    albumWrapper.isUserLoggedIn = jest.fn();
    let mockEvent = {preventDefault: jest.fn()}

    albumWrapper.find('button').at(0).simulate('click', mockEvent);

    expect(albumWrapper.isUserLoggedIn).toHaveBeenCalledWith(mockEvent);
  });

  it('should call isUserLoggedIn on a button click', () => {

    albumWrapper.isUserLoggedIn = jest.fn();
    let mockEvent = { preventDefault: jest.fn() }

    albumWrapper.find('button').at(1).simulate('click', mockEvent);

    expect(albumWrapper.isUserLoggedIn).toHaveBeenCalledWith(mockEvent);
  });
});

describe('mapStateToProps', () => {
  it('should return an object with the favorites array', () => {

    const mockState = {
      albums: [],
      user: {id: 1, name: 'Alan', email: 'alan@turing.io'},
      error: '',
      favorites: [
        {
          album_id: 626204707,
          artist_name: 'Beyoncé',
          album_name: '4 (Expanded Edition)',
          price: 11.99,
          artwork_url: 'https://is2-ssl.mzstatic.com/image/thumb/Music6/v4/17/84/3a/17843a6d-1f2b-7e1e-a39f-3ff865110993/source/100x100bb.jpg',
          release_date: '2013-12-13T08:00:00Z',
          content_advisory_rating: 'Explicit',
          primary_genre_name: 'Pop',
        }
      ]
    }

    const expected = {
      favorites: [
        {
          album_id: 626204707,
          artist_name: 'Beyoncé',
          album_name: '4 (Expanded Edition)',
          price: 11.99,
          artwork_url: 'https://is2-ssl.mzstatic.com/image/thumb/Music6/v4/17/84/3a/17843a6d-1f2b-7e1e-a39f-3ff865110993/source/100x100bb.jpg',
          release_date: '2013-12-13T08:00:00Z',
          content_advisory_rating: 'Explicit',
          primary_genre_name: 'Pop',
        }
      ]
    }

    const mappedProps = mapStateToProps(mockState);

    expect(mappedProps).toEqual(expected)
  });
});

describe('mapDispatchToProps', () => {
  
})