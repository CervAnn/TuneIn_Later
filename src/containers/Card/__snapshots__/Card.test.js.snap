// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot with all data passed in correctly 1`] = `
<article
  className="favoriteCard"
>
  <Link
    to="/albums/626204707"
  >
    <img
      alt="Album cover art"
      id="standardCard"
      src="https://is2-ssl.mzstatic.com/image/thumb/Music6/v4/17/84/3a/17843a6d-1f2b-7e1e-a39f-3ff865110993/source/100x100bb.jpg"
    />
  </Link>
  <h3>
    4 (Expanded Edition)
  </h3>
  <article
    className="favorite"
  >
    <img
      alt=""
      id="isFavorite"
      onClick={[Function]}
      src={
        Object {
          "ReactComponent": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "__esModule": true,
          "default": "isFavorite.svg",
        }
      }
    />
    <button
      id="isFavorite-button"
      onClick={[Function]}
    >
      Remove Favorite
    </button>
  </article>
</article>
`;

exports[`Card should match the snapshot with all data passed in correctly 2`] = `
<article
  className="favoriteCard"
>
  <Link
    to="/favorites/626204707"
  >
    <img
      alt="Album cover art"
      id="standardCard"
      src="https://is2-ssl.mzstatic.com/image/thumb/Music6/v4/17/84/3a/17843a6d-1f2b-7e1e-a39f-3ff865110993/source/100x100bb.jpg"
    />
  </Link>
  <h3>
    4 (Expanded Edition)
  </h3>
  <article
    className="favorite"
  >
    <img
      alt=""
      id="isFavorite"
      onClick={[Function]}
      src={
        Object {
          "ReactComponent": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "__esModule": true,
          "default": "isFavorite.svg",
        }
      }
    />
    <button
      id="isFavorite-button"
      onClick={[Function]}
    >
      Remove Favorite
    </button>
  </article>
</article>
`;

exports[`Card should match the snapshot with all data passed in correctly 3`] = `
<article
  className="Card"
>
  <Link
    to="/favorites/626204708"
  >
    <img
      alt="Album cover art"
      id="standardCard"
      src="https://is2-ssl.mzstatic.com/image/thumb/Music6/v4/17/84/3a/17843a6d-1f2b-7e1e-a39f-3ff865110993/source/100x100bb.jpg"
    />
  </Link>
  <h3>
    4 (Expanded Edition)
  </h3>
  <article
    className="favorite"
  >
    <img
      alt=""
      id="notFavorite"
      onClick={[Function]}
      src={
        Object {
          "ReactComponent": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "__esModule": true,
          "default": "notFavorite.svg",
        }
      }
    />
    <button
      id="notFavorite-button"
      onClick={[Function]}
    >
      Add Favorite
    </button>
  </article>
</article>
`;
